client/v2/tx/config.go

```diff
@@ -0,0 +1,231 @@
+	if c.AddressCodec == nil {
+		return errors.New("address codec cannot be nil")
+	}
+	if c.Cdc == nil {
+		return errors.New("codec cannot be nil")
+	}
+	if c.ValidatorAddressCodec == nil {
+		return errors.New("validator address codec cannot be nil")
+	}
+
+	// set default signModes
+	if len(c.EnablesSignModes) == 0 {
+		c.EnablesSignModes = defaultEnabledSignModes
+	}
+	return nil
+}
+
+type txConfig struct {
+	TxBuilderProvider
+	TxEncodingConfig
+	TxSigningConfig
+}
+
+func NewTxConfig(options ConfigOptions) (TxConfig, error) {
+	err := options.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	signingCtx, err := newDefaultTxSigningConfig(options)
+	if err != nil {
+		return nil, err
+	}
+
+	if options.Decoder == nil {
+		options.Decoder, err = txdecode.NewDecoder(txdecode.Options{SigningContext: signingCtx.SigningContext()})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &txConfig{
+		TxBuilderProvider: NewBuilderProvider(options.AddressCodec, options.Decoder, options.Cdc),
+		TxEncodingConfig:  defaultEncodingConfig{},
+		TxSigningConfig:   signingCtx,
+	}, nil
+}
+
+type defaultEncodingConfig struct{}
+
+func (t defaultEncodingConfig) TxEncoder() txApiEncoder {
+	return txEncoder
+}
+
+func (t defaultEncodingConfig) TxDecoder() txApiDecoder {
+	return txDecoder
+}
+
+func (t defaultEncodingConfig) TxJSONEncoder() txApiEncoder {
+	return txJsonEncoder
+}
+
+func (t defaultEncodingConfig) TxJSONDecoder() txApiDecoder {
+	return txJsonDecoder
+}
+
+type defaultTxSigningConfig struct {
+	signingCtx *signing.Context
+	handlerMap *signing.HandlerMap
+}
+
+func newDefaultTxSigningConfig(opts ConfigOptions) (*defaultTxSigningConfig, error) {
+	signingCtx, err := newSigningContext(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	handlerMap, err := newHandlerMap(opts, signingCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &defaultTxSigningConfig{
+		signingCtx: signingCtx,
+		handlerMap: handlerMap,
+	}, nil
+}
+
+func (t defaultTxSigningConfig) SignModeHandler() *signing.HandlerMap {
+	return t.handlerMap
+}
+
+func (t defaultTxSigningConfig) SigningContext() *signing.Context {
+	return t.signingCtx
+}
+
+func (t defaultTxSigningConfig) MarshalSignatureJSON(signatures []Signature) ([]byte, error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (t defaultTxSigningConfig) UnmarshalSignatureJSON(bytes []byte) ([]Signature, error) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func newSigningContext(opts ConfigOptions) (*signing.Context, error) {
+	return signing.NewContext(signing.Options{
+		FileResolver:          opts.FileResolver,
+		TypeResolver:          opts.TypeResolver,
+		AddressCodec:          opts.AddressCodec,
+		ValidatorAddressCodec: opts.ValidatorAddressCodec,
+		CustomGetSigners:      opts.CustomGetSigner,
+		MaxRecursionDepth:     opts.MaxRecursionDepth,
+	})
+}
+
+func newHandlerMap(opts ConfigOptions, signingCtx *signing.Context) (*signing.HandlerMap, error) {
```

Comment:
And this one also could be moved to client/offchain to avoid having too much signing decisions here

Reply:
ditto